Corrfunc/mocks/DDsmu_mocks.py

```diff
@@ -0,0 +1,304 @@
+       can be used to compute the actual :math:`\\xi(s, \mu)` by combining with
+       (DR, RR) counts.
+
+    api_time: float, optional
+       Only returned if ``c_api_timer`` is set.  ``api_time`` measures only the time
+       spent within the C library and ignores all python overhead.
+    """
+    try:
+        from Corrfunc._countpairs_mocks import countpairs_s_mu_mocks as\
+            DDsmu_extn
+    except ImportError:
+        msg = "Could not import the C extension for the on-sky"\
+              "pair counter."
+        raise ImportError(msg)
+
+    import numpy as np
+    from Corrfunc.utils import translate_isa_string_to_enum, fix_ra_dec,\
+        return_file_with_rbins
+    from future.utils import bytes_to_native_str
+
+    # Broadcast scalar weights to arrays
+    if weights1 is not None:
+        weights1 = np.atleast_1d(weights1)
+    if weights2 is not None:
+        weights2 = np.atleast_1d(weights2)
+
+    if not autocorr:
+        if RA2 is None or DEC2 is None or CZ2 is None:
+            msg = "Must pass valid arrays for RA2/DEC2/CZ2 for "\
+                  "computing cross-correlation"
+            raise ValueError(msg)
+
+        # If only one set of points has weights, set the other to uniform weights
+        if weights1 is None and weights2 is not None:
+            weights1 = np.ones_like(weights2)
+        if weights2 is None and weights1 is not None:
+            weights2 = np.ones_like(weights1)
+
+    else:
+        RA2 = np.empty(1)
+        DEC2 = np.empty(1)
+        CZ2 = np.empty(1)
+
+    fix_ra_dec(RA1, DEC1)
+    if autocorr == 0:
+        fix_ra_dec(RA2, DEC2)
+
+    # Passing None parameters breaks the parsing code, so avoid this
+    kwargs = {}
+    for k in ['weights1', 'weights2', 'weight_type', 'RA2', 'DEC2', 'CZ2']:
+        v = locals()[k]
+        if v is not None:
+            kwargs[k] = v
+
+    integer_isa = translate_isa_string_to_enum(isa)
+    sbinfile, delete_after_use = return_file_with_rbins(binfile)
+    extn_results, api_time = DDsmu_extn(autocorr, cosmology, nthreads,
+                                         nmu_bins, mu_max, sbinfile,
+                                         RA1, DEC1, CZ1,
+                                         is_comoving_dist=is_comoving_dist,
+                                         verbose=verbose,
+                                         output_savg=output_savg,
+                                         fast_divide=fast_divide,
+                                         xbin_refine_factor=xbin_refine_factor,
+                                         ybin_refine_factor=ybin_refine_factor,
+                                         zbin_refine_factor=zbin_refine_factor,
+                                         max_cells_per_dim=max_cells_per_dim,
+                                         c_api_timer=c_api_timer,
+                                         isa=integer_isa, **kwargs)
+    if extn_results is None:
```

Comment:
Sometimes DDsmu_extn returns None instead of a tuple, then line 259 fails before 272 raises the error.

Reply:
Do you know when that happens? Usually, `None` is returned when an error occurs

Reply:
That's exactly what I mean. If None is returned, then we can't unpack it to a 2-tuple. It happens for example if the input data is all identical.

Reply:
BTW the error message when input data is all identical is not very useful. It is currently something like `(ix = -214xxxxx is not > 0)`.

Also Python do not see these error messages. They are printed to the stderr/stdout. It probably requires a major change to the code to add the ability to raise them as proper Python exceptions.

Reply:
Here is my bug reproducer:

```
(test) [yfeng1@waterfall tmp]$ cat el.py
from nbodykit.lab import *
import numpy as np

################### INPUT ###################

# selections                                                            
R_min = 3818.0
R_max = 3986.5

f_random = 10  # ratio between the number of random points and the number of galaxies selected                                                 
Ngal_select = 1000000


gal_pos = np.random.uniform(size=(3,3)) - 0.4
gal_pos[...] = 0

print('positions read')
cat = ArrayCatalog({'Position' : gal_pos},
                                      BoxSize = np.array([1e0,1e0,1e0]),

)

#print(cat['Position'].shape)

array_r = np.linspace(0.0,0.1,10)
DD = paircount.SimulationBoxPairCount(mode='2d',source1 = cat, 
                                      source2 = cat,
                                      periodic=False,
                                      BoxSize = np.array([1e0,1e0,1e0]),
                                      redges = array_r,
                                      Nmu = 10)
```

Reply:
We can create a `all data are identical` check into the python wrapper. Otherwise, the error handling will have to be re-written entirely. 

The particular error you are seeing is probably from attempting to compute a grid cell index when the depth of data is `0` and that leads to integer overflow.

I am confused about python not seeing the error messages - the messages printed to stderr show up both in ipython and python for me. Does it not for you?

Reply:
I mean they are not represented as any entities in the Python interpretor instance.

Reply:
Ahh, I think I see the easy fix. The call has to be wrapped in a try/catch, and then the `RuntimeError` can be raised. 

The correct option -- to propagate errors -- is far too time-consuming.

Reply:
Yes. I marginally agree with you; but maybe not that bad.


You probably already have a macro or a function for printing errors. So what about strcpy the last error to a static variable that can be fetched with the API?

Reply:
Since the errors can occur at many different levels of the code, where most of the functions do not have the function signature to handle a separate argument. And I refuse to go the route of global error message and error status. If the error is at the python wrapper level, then what you suggest is already somewhat implemented. 

(Error-handling is vastly different between `python2` and `python3`, and the C interface for the extension is already sub-optimal with the use of `static` variables for error-handling. So I will prefer any specific error-checks to be handled at the `python` level rather than the C-level)

Reply:
I suspect one of the original motivation of exception handling is to defer the validation to closer to the consumer of the data (whom by definition has an accurate understanding of the schema). Of course in scientific applications the schema is so clear it makes sense to validate early.

I agree it's a bad idea to add error handling as an after-thought; whatever you do it tasts like wax.Will you add the check in the Python interface?